datalad/local/foreach.py

```diff
@@ -0,0 +1,275 @@
+        subdatasets_only=Parameter(
+            args=("-s", "--subdatasets-only"),
+            action="store_true",
+            doc="""whether to exclude top level dataset."""),
+        passthrough=Parameter(  # TODO  could be of use for `run` as well
+            args=("-p", "--passthrough"),
+            action="store_true",
+            doc="""For command line commands, pass-through their output to 
+            the screen instead of capturing/returning as part of the result records."""),
+        jobs=jobs_opt,
+        # TODO: might want explicit option to either worry about 'safe_to_consume' setting for parallel
+        # For now - always safe
+    )
+
+    @staticmethod
+    @datasetmethod(name='foreach')
+    @eval_results
+    def __call__(
+            cmd=None,
+            python=False,
+            dataset=None,
+            fulfilled=None,
+            recursive=False,
+            recursion_limit=None,
+            # contains=None,
+            bottomup=False,
+            subdatasets_only=False,
+            passthrough=False,
+            jobs=None
+            ):
+        if not cmd:
+            lgr.warning("No command given")
+            return
+        if python and len(cmd) > 1:
+            # yoh decided to avoid unnecessary complication/inhomogeneity with support
+            # of multiple Python commands for now
+            raise ValueError(f"Please provide a single Python expression. Got {len(cmd)}: {cmd!r}")
+        refds = require_dataset(
+            dataset, check_installed=True, purpose='foreach execution')
+        pwd = getpwd()  # Note: 'run' has some more elaborate logic for this
+        subdatasets_it = refds.subdatasets(
+            fulfilled=fulfilled, recursive=recursive, recursion_limit=recursion_limit,
+            bottomup=bottomup,
+            result_xfm='paths'
+        )
+
+        if subdatasets_only:
+            datasets_it = subdatasets_it
+        else:
+            if bottomup:
+                datasets_it = chain(subdatasets_it, [refds.path])
+            else:
+                datasets_it = chain([refds.path], subdatasets_it)
+
+        if not python:
+            protocol = NoCapture if passthrough else StdOutErrCapture
+
+        def run_cmd(dspath):
+            ds = Dataset(dspath)
+            status_rec = get_status_dict(
+                'foreach',
+                ds=ds,
+                path=ds.path,
+                command=cmd
+            )
+            if not ds.is_installed():
+                yield dict(
+                    status_rec,
+                    status="impossible",
+                    message="not installed"
+                )
+                return
+            # For consistent environment (Python) and formatting (command) similar to `run` one
+            # But for Python command we provide actual ds and refds not paths
+            placeholders = dict(
+                pwd=pwd,
```

Comment:
At least for the non-python command, I'd find it confusing that `pwd` doesn't correspond to the cwd for execution.

Reply:
yeah, I think I am of the same opinion now -- will make them consistently `chpwd` although (later) we might like to add an option `--chpwd=True|False` to facilitate possible cases where it is desired to operate from current directory?

Reply:
I think I misunderstood/misplaced my comment -- I thought about actual directory under which we run.  @mih raised comment about that [below](https://github.com/datalad/datalad/pull/5517#discussion_r598808830) so I will move my answer there

Reply:
@kyleam re `pwd`.  
- I do not see a value ATM to make `pwd` correspond to cwd for execution since that one would be identical to `{ds.path}`
- I think it is good to have it consistent with `run` which seems to have it to correspond to the current directory from which `run` is initiated: ""{pwd}" will be replaced with the full path of the current working directory. ":
```shell
(git)lena:~datalad/datalad-master[nf-foreach]git
$> datalad run -d ~/datalad/openneuro --explicit echo '{pwd}'            
[INFO   ] == Command start (output follows) ===== 
/home/yoh/proj/datalad/datalad-master
```

Reply:
> with `run` which seems to have it to correspond to the current directory from which `run` is initiated

Unless a dataset instance is given, the directory in which `run` is called and the execution directory are one and the same.  The `pwd` passed as `WitlessRunner`s `cwd` and the placeholder are the same value.

Reply:
> Unless a dataset instance is given, the directory in which `run` is called and the execution directory are one and the same

you might be in a subdirectory as well:
```shell
(git)lena:~datalad/datalad-master[nf-foreach]git
$> datalad -C datalad run --explicit 'echo {pwd} {dspath}' 
[INFO   ] == Command start (output follows) ===== 
/home/yoh/proj/datalad/datalad-master/datalad /home/yoh/proj/datalad/datalad-master
[INFO   ] == Command exit (modification check follows) ===== 
```
and since `foreach` (if `chdir`ing - yet to add option) would run at the top directory of a (sub)dataset, `{ds.path}` would provide that path, but without `{pwd}` there would be no way to perform any operation which needs `cwd` without relying on some env var like `PWD` which might be set or not.
So I still do not see a point of 1. breaking analogie to `run`; 2. making `{pwd} == {ds.path}` and thus a duplicate; 3. removing possibility to have access to the initial directory from which `foreach` is executed.

Reply:
> > Unless a dataset instance is given, the directory in which `run` is called and the execution directory are one and the same
> 
> you might be in a subdirectory as well

Yes, you can use `-C` to, in the option's own words, "run as if datalad was started in \<path> instead of the current working directory", and that of course affects run's pwd/the directory of execution.  Sorry, I'm not following what your point is by mentioning that, though.  

> 1. breaking analogie to run

Again, `run`s `pwd` is the execution directory for the command.  And, if a dataset instance is passed, that corresponds to `ds.path`.   If it's not, it's `getpwd()` (either `ds.path` or a subdirectory).  Either way, it's the actual `cwd` value passed to `WitlessRunner`.

I see a discrepancy with `run`, not an analogy. But...

> 2. making `{pwd} == {ds.path}` and thus a duplicate; 3. removing possibility to have access to the initial directory from which `foreach` is executed.

... fair enough. Perhaps it's worth coming up with another name, maybe `{pwd_initial}`?

Reply:
After adding `--chpwd` I am now tending to agree with you @kyleam :  it becomes odd to have `--chpwd true` and then `{pwd}` still be the original one.  I also had difficult time to get the "effective" pwd while testing it out with or without --chpwd ;)
- I also dislike naming of `--chpwd` and tend to think about following @mih mantra of no boolean flags: we might want eventually to add `relds` or alike setting to invoke in the matching subdirectory of a dataset (e.g. when I am under some `.datalad/metadata/objects` or alike).  So may be it should become `--pwd ds,cwd` or alike?
- with the above -- may be `{cwd}` for the `{pwd_initial}` as for "current working directory"?  It might be a bit confusing since generally cwd and pwd are for the same thing, but IMHO "nothing the documentation can't address" ;)

WDYT on `--pwd` and `{cwd}`?

Reply:
> WDYT on `--pwd` and `{cwd}`?

I'm a bit turned around, but that sounds okay to me.  Given your points 2 and 3 [above][1], I'm also okay leaving it at `pwd`, though I think using a different name is probably better.  Either way, I agree that this area is confusing, and we'll need some good documentation :)

[1]: https://github.com/datalad/datalad/pull/5517#discussion_r626916216

Reply:
pushed changes to have it 
```
  --chpwd {ds|pwd}      'ds' will change working directory to the top of the corresponding dataset. With 'pwd' no change of
                        working directory will happen. Note that for Python commands, due to use of threads, we do not
                        allow chdir=ds to be used with jobs > 1. Hint: use 'ds' and 'refds' objects' methods to execute
                        commands in the context of those datasets. Constraints: value must be one of ('ds', 'pwd')
                        [Default: 'ds']
```